validMoves and boardAfter

vincentfrochot opened the pull request.

vincentfrochot committed:
feat: moves of 2 marbles

JamesHeppell commented on src/main/scala/abalone/Board.scala:
```diff
@@ -20,6 +20,11 @@ case class Board(
 
   def piecesOf(player: Player): PieceMap = pieces.filter(_._2.is(player))
 
+  def isEmptySquare(pos: Option[Pos]): Boolean = pos match {
```
when you have a match with only 2 cases, a Some and None, then you can use a map or fold depending on the output.  Here I would use a fold, e.g. 
`pos.fold(false)(!this.pieces.contains(_))`

vincentfrochot replied:
thank you

vincentfrochot committed:
chore: codestyle

JamesHeppell commented on src/main/scala/abalone/Pos.scala:
```diff
@@ -190,8 +190,38 @@ case class Pos private (index: Int) extends AnyVal {
+  }
+  def dir(pos: Pos): Option[String] = {
+    (pos.file.index - this.file.index, pos.rank.index - this.rank.index) match {
+      case (0, 1) => Some("upLeft")
```
can you double check the coordinates it doesn't seem right to me, right and left swapping whether they are negative or not.

JamesHeppell replied:
disregard this, it looks correct upon me looking at the diagram, just felt wrong but that the new coordinate system for you.

vincentfrochot replied:
I do not understand I think
I tried this in console :
```
scala> strategygames.abalone.Pos.A1.dir(strategygames.abalone.Pos.B1)
val res2: Option[String] = Some(right)

scala> strategygames.abalone.Pos.B1.dir(strategygames.abalone.Pos.A1)
val res3: Option[String] = Some(left)
```

vincentfrochot replied:
ah ok :+1:

JamesHeppell commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,6 +103,144 @@ abstract class Variant private[variant] (
+    }.toMap
+
+  // def validMovesOf3(situation: Situation, @nowarn validMoves2: Map[Pos, List[(String, Move)]]): Map[Pos, List[(String, Move)]] =
+  //   situation.board.piecesOf(situation.player).flatMap {
```
as above, if you are reusing the same code we should refactor into val, e.g. 
```
{
val turnPieces = situation.board.piecesOf(situation.player)
code to return...
}
```

vincentfrochot replied:
do not review movesOf3 it is not ready, should have just not committed sorry

vincentfrochot replied:
but so you mean I should put `val turnPieces = situation.board.piecesOf(situation.player)` into e.g. validMoves() so it can be reused by validMovesOf1 and validMovesOf2 at least for example ?

msomatt commented on src/main/scala/abalone/Pos.scala:
```diff
@@ -190,8 +190,36 @@ case class Pos private (index: Int) extends AnyVal {
+      case Some("downRight") => this.downRight
+      case Some("upRight") => this.upRight
+      case _ => None
+    }
```
I wonder if you can reuse the idea of `Direction` which is in the chess game logic, instead of having strings for each direction?

vincentfrochot replied:
actually in abalone package.scala we have the same :

```
  type Direction  = Pos => Option[Pos]
  type Directions = List[Direction]
```

Do you want I start having a look at how to use this properly then ?

JamesHeppell replied:
sure, using a type with the 6 options instead of a string, is safer (typesafety).

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,6 +103,144 @@ abstract class Variant private[variant] (
+        Map(
+          pos -> pos.neighbours.flatMap {
+            case Some(neighbour) if(situation.board.piecesOf(situation.player).contains(neighbour)) => {
+              neighbour.neighbours.flatMap {
```
it would be better if we could just get the neighbour of `neighbour` only in the direction of pos->neighbour? all of the other neighbour of neighbours are just irrelevant?

vincentfrochot replied:
so I think we want to consider the positions where ab could land, displayed as uppercase letter in the situation below
```
   *  *  A  B  *
 *  *  a  b  C  *
*  *  *  E  D  *  *  
```
?

msomatt replied:
But ab can only push into C right? theres no valid move ever for ab to interact with A, B, D or E, so why even call a function that returns A,B,C,D,E?

vincentfrochot replied:
In Abalone there are 3 kinds of moves.
  Let's use the top 2 rows of a board to illustrate (numbers are empty squares, Z is an opponent marble)

        Z a b c 7
       1 2 3 4 5 6
    
    - line moves are marbles moving in line to an empty square :
       'a' to '7' would move three marbles to the right. This is the same as if 'a' jumped over 'b' and 'c' to land on '7'.
       'c' to '7' would move one marble to the right.
    - pushes are line moves targeting a square hosting an opponent marble
       They will be processed as two line moves (one jump per player).
       'c' to 'Z' and 'b' to 'Z' are pushes to the left
    - side moves can only be described starting from the right origin (figure out the longest diagonal) :
       'a' to '5' is a downRight side move of three marbles.
       'c' to '2' is a downLeft side move of three marbles.
       'a' to '4' is a downRight side move of two marbles (only 'a' and 'b' would move).
       'c' to '3' is a downLeft side move of two marbles (only 'c' and 'b' would move)
       'a' to '3' or 'c' to '4' are line moves of a single marble

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,6 +103,144 @@ abstract class Variant private[variant] (
+              neighbour.neighbours.flatMap {
+                case Some(neighbourOfNeighbour)
+                  if (situation.board.isEmptySquare(Some(neighbourOfNeighbour))
+                  && (pos.isInLine(neighbour, neighbourOfNeighbour) || situation.board.isEmptySquare(pos.dir(neighbour.dir(neighbourOfNeighbour))))
```
I'm not sure i understand why we do the second half of this line (`situation.board.isEmptySquare(pos.dir(neighbour.dir(neighbourOfNeighbour)))`)

vincentfrochot replied:
it is to handle that case : 
```
   *  *  Z  *  *
 *  *  a  b  *  *
*  *  *  *  *  *  *  
```
dir can be applied to b but not applied to a

vincentfrochot replied:
(if we move upRight)

vincentfrochot replied:
with `if (situation.board.isEmptySquare(Some(neighbourOfNeighbour))`
neighbourOfNeighbour could take the value of Z

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,6 +103,144 @@ abstract class Variant private[variant] (
+                case Some(neighbourOfNeighbour)
+                  if (situation.board.isEmptySquare(Some(neighbourOfNeighbour))
+                  && (pos.isInLine(neighbour, neighbourOfNeighbour) || situation.board.isEmptySquare(pos.dir(neighbour.dir(neighbourOfNeighbour))))
+                  && !validMoves1.get(pos).toList.flatMap(_.map(_.toUci)).exists(m => m.toString == s"Move(${pos},${neighbourOfNeighbour},None)")) =>
```
i think there must be a way not to need to look at `validMoves1` here. Surely by checking the whole neighbour of neighbours we are doing a search that involves three `Pos` here, and so we can't be generating a move of 1 in this code?

JamesHeppell replied:
you could probably filter our moves where the origin and dest are neighbours as no moves of two can have that form?

msomatt replied:
yes, i don't think there is a need to have to check `validMovesOf1` in `validMovesOf2` and it would be much more efficient if we didn't have to pass in a parameter `validMovesOf1`.

JamesHeppell commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,6 +103,144 @@ abstract class Variant private[variant] (
         )
     }.toMap
 
+  def validMovesOf2(situation: Situation, validMoves1: Map[Pos, List[Move]]): Map[Pos, List[(String, Move)]] =
```
some comments to understand the cases could be helpful, it was a little tricky to follow through first time.

JamesHeppell replied:
but given the tests work, its not strictly necessary.

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,6 +103,144 @@ abstract class Variant private[variant] (
+                  && (pos.isInLine(neighbour, neighbourOfNeighbour) || situation.board.isEmptySquare(pos.dir(neighbour.dir(neighbourOfNeighbour))))
+                  && !validMoves1.get(pos).toList.flatMap(_.map(_.toUci)).exists(m => m.toString == s"Move(${pos},${neighbourOfNeighbour},None)")) =>
+                    Some( ("nopush", Move(piece, pos, neighbourOfNeighbour, situation, boardAfter(situation, neighbour, neighbourOfNeighbour), false)) )
+                case Some(neighbourOfNeighbour)
```
from here on, this function makes sense :)

vincentfrochot committed:
feat: generate movesOf2 based on pairs of marbles

JamesHeppell commented on src/main/scala/abalone/Pos.scala:
```diff
@@ -190,7 +190,41 @@ case class Pos private (index: Int) extends AnyVal {
-  def neighbours: List[Option[Pos]] = List(left, downLeft, upLeft, right, downRight, upRight)
+
+  def neighbours: List[Option[Pos]] = List(left, upLeft, upRight, right, downRight, downLeft)
+  def sideMovesDirsFromDir(dir: Option[String]): (Option[Pos], Option[Pos]) = Map(
```
Same as below, its unnecessary that you are creating a map and then getting an item from it, just you a match on the dir and have a default case returning none.
same for function below.

JamesHeppell commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -90,18 +101,82 @@ abstract class Variant private[variant] (
+  def turnPieces(s: Situation): PieceMap = s.board.piecesOf(s.player)
+
+  def validMovesOf1(s: Situation): Map[Pos, List[Move]] =
+    this.turnPieces(s).flatMap {
```
pretty sure you dont need `this.` also same down below.

JamesHeppell commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -90,18 +101,82 @@ abstract class Variant private[variant] (
     }.toMap
 
+  def validMovesOf2(s: Situation): Map[Pos, List[(String, Move)]] = {
+    def generateMove(orig: Pos, dest: Pos, category: String) = category match {
```
just a comment for future, given the cases are very similar (and you only have two) you could remove and have an if statement inside the last input of Move (capture) which would shorten the code. But its fine the way you have written it.

vincentfrochot committed:
feat: generate validMoves

vincentfrochot commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,17 +102,80 @@ abstract class Variant private[variant] (
-  def validSideMoves(@nowarn situation: Situation):  Map[Pos, List[Move]] = {
-    Map()
-  }
+  def validMovesOf2And3(situation: Situation): Map[Pos, List[(String, Move)]] = {
```
not sure what better name we could pick

JamesHeppell replied:
name is clear enough

vincentfrochot commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,17 +102,80 @@ abstract class Variant private[variant] (
+        if (lineOfMarbles.size == 3) generateSideMoves(List(lineOfMarbles(0), lineOfMarbles(1)), direction)
+        else None,
+        if (!lineOfMarbles.map(canLeftSideMove(situation, _, direction)).contains(false))
+          generateMove(lineOfMarbles(0), lineOfMarbles.last.sideMovesDirsFromDir(direction)._1.get, "side")
```
I'm still having .get

msomatt replied:
If ever you want to use `.get` this means that you might have `None`, and you'll want to handle the case where you have `None`. Writing `.get` allows you to compile, but if this ever resolves to `None` at run time you will get a run time error which you haven't handled.
So instead you need to `map` on the `Option` which allows you to have the concrete value inside the `map` and ensures this code never runs if you have `None`. 
I also don't like `last` or `lineOfMarbles(0)` because these also fail at run time on an empty list (I mentioned `dropRight` above and this is safe because `List().dropRight(1)` == `List()`.
So I would use `lineOfMarbles.headOption` which gives you an `Option` of the 0th element and `None` if `lineOfMarbles.isEmpty`. For `.last` I use `.reverse.headOption`.

This means you want to use a combination of map and flatMap chained together to call `generateMove` safely.
Consider this simplified example:
```
scala> def myFunc(a: Int, b: Int): Option[Int] = Some(a+b) //this is representing generateMove
def myFunc(a: Int, b: Int): Option[Int]

scala> val myList = List(11,12,13)
val myList: List[Int] = List(11, 12, 13)

scala> def intToOpt(i: Int): Option[Int] = Some(i) //this is representing sideMovesDirsFromDir(direction)._1
def intToOpt(i: Int): Option[Int]

scala> myFunc(myList(0), intToOpt(myList.last).get) //this is what you are currently doing
val res11: Option[Int] = Some(24)

scala> myList.headOption.flatMap(h => myFunc(h, intToOpt(myList.last).get)) //handling lineOfMarbles(0) safely
val res15: Option[Int] = Some(24)

scala> myList.headOption.flatMap{h => myList.reverse.headOption.flatMap{t => myFunc(h, intToOpt(t).get)}} //handling lineOfMarbles.last safely
val res16: Option[Int] = Some(24)

scala> myList.headOption.flatMap{h => myList.reverse.headOption.flatMap{t => intToOpt(t).flatMap{t2 => myFunc(h, t2)}}} //everything safe now, intToOpt handled.
val res17: Option[Int] = Some(24)
```
Now if `myList` was empty this is the difference:
```
scala> val myEmptyList: List[Int] = List()
val myEmptyList: List[Int] = List()

scala> myEmptyList.headOption.flatMap{h => myEmptyList.reverse.headOption.flatMap{t => intToOpt(t).flatMap{t2 => myFunc(h, t2)}}}
val res18: Option[Int] = None

scala> myFunc(myEmptyList(0), intToOpt(myEmptyList.last).get)
java.lang.IndexOutOfBoundsException: 0
  at scala.collection.LinearSeqOps.apply(LinearSeq.scala:117)
  at scala.collection.LinearSeqOps.apply$(LinearSeq.scala:114)
  at scala.collection.immutable.List.apply(List.scala:79)
  ... 59 elided
```

vincentfrochot commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,17 +102,80 @@ abstract class Variant private[variant] (
+          generateMove(lineOfMarbles(0), lineOfMarbles.last.sideMovesDirsFromDir(direction)._1.get, "side")
+        else None,
+        if (!lineOfMarbles.map(canRightSideMove(situation, _, direction)).contains(false))
+          generateMove(lineOfMarbles(0), lineOfMarbles.last.sideMovesDirsFromDir(direction)._2.get, "side")
```
I'm still having .get

vincentfrochot commented on src/main/scala/abalone/Pos.scala:
```diff
@@ -190,7 +190,43 @@ case class Pos private (index: Int) extends AnyVal {
+
+  // NOTE - *neighbourhood
+  // these below only work for neighbour pos but that's probably fine as in Abalone we only move to (potentially extended) neighbourhood
+  def dir(dir: Option[String]): Option[Pos]  = dir match {
```
still using String for direction...

msomatt replied:
So did you work out a way to reuse `Direction` here? I see `Directions` used in other places in the code

vincentfrochot committed:
chore: move sideMoves methods under parent one

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -107,13 +107,19 @@ abstract class Variant private[variant] (
+        pos.sideMovesDirsFromDir(direction)._2.fold(false)(p => situation.board.isEmptySquare(Some(p)))
+
       List(
         if (lineOfMarbles.size == 3) generateSideMoves(List(lineOfMarbles(0), lineOfMarbles(1)), direction)
```
I don't understand this line. Why do we trigger a special check in the case we have 3 marbles? Why does it only use the first two? If we are looking at pairs from three, do we need to also check `List(lineOfMarbles(1), lineOfMarbles(2))`?
Also `List(lineOfMarbles(0), lineOfMarbles(1)` == `lineOfMarbles.dropRight(1)` if `lineOfMarbles.size == 3`.

vincentfrochot replied:
for a side move to be possible, all the elements in the line need to be able to move in the direction specified.
we then generate a side move of 3 marbles with pos and 3rdMarblePos.dir(direction)
for side moves of 2 marbles, we generate it with pos and 2ndMarblePos.dir(direction)

vincentfrochot replied:
so if the size is 3, I make a call to itself but with the shorter list of 2 elements to generate side moves of 2 marbles too

vincentfrochot replied:
If we are looking at pairs from three, do we need to also check List(lineOfMarbles(1), lineOfMarbles(2))?
=> no, these side moves will be generated later (or have already been) because it will be triggered from another initial pos

vincentfrochot replied:
I'll change `generateSideMoves(List(lineOfMarbles(0), lineOfMarbles(1)), direction)` to `generateSideMoves(lineOfMarbles.dropRight(1))` thank you

msomatt replied:
Ok I think I understand why we dont need to check the other pair in the list of three.
However, are we not already checking groups of two in another way? (I've not read allll the code to be sure). How would a group of two that is not part of a subgroup of 3 be handled?
Or is the groups of three marbles (`lineOfMarbles`) every group of three Pos?

vincentfrochot replied:
so when we enter the moves generator as (A, B, right) it will generate the side moves of ABC
(and when we enter as C, B, left), the ones for CBA

msomatt replied:
So the way I read it is, if we have A B C as three consecutive marbles of the same colour we generate the side moves of A B as well as A B C. The side moves of B C will be covered when handling C B A later.
What I don't get is when would you cover the side moves of A B if C wasn't a marble of the same colour, and it was either an empty pos or an opponents marble?

vincentfrochot replied:
it is more : 
we will generate the side moves of A B since we know it exists as a valid pair.
Pushing into that same direction of A towards B, if a C exists we will also generate the side moves of A B C

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -107,13 +107,19 @@ abstract class Variant private[variant] (
         if (lineOfMarbles.size == 3) generateSideMoves(List(lineOfMarbles(0), lineOfMarbles(1)), direction)
         else None,
-        if (!lineOfMarbles.map(canLeftSideMove(situation, _, direction)).contains(false))
+        if (!lineOfMarbles.map(canLeftSideMove(_)).contains(false))
```
can remove `(_)` from here

vincentfrochot committed:
chore: use dropRight on List

vincentfrochot committed:
chore: map(func) equals map(func(_))

vincentfrochot commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,17 +102,86 @@ abstract class Variant private[variant] (
+          case (thirdSquareInLine) => {
+            if (situation.board.isEmptySquare(Some(thirdSquareInLine))) // xx.
+              generateMove(pos, thirdSquareInLine, "line")
+            else if (situation.board.piecesOf(!situation.player).contains(thirdSquareInLine)) // xxo
```
here I think I should instead have defined `val opponentPieces = situation.board.piecesOf(!situation.player)` because I reuse it several times. Same for "myPieces" probably

vincentfrochot committed:
chore: situation.board.piecesOf(x) as val

vincentfrochot commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,17 +102,88 @@ abstract class Variant private[variant] (
+    turnPieces(situation).flatMap {
+      case (pos, _) =>
+        Map(pos -> pos.neighbours.collect {
+          case Some(neighbour) if situation.board.piecesOf(situation.player).contains(neighbour) => (pos, neighbour, pos.dir(neighbour))
```
will swap `situation.board.piecesOf(situation.player).` with the new `activePlayerPieces`

JamesHeppell commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,17 +102,88 @@ abstract class Variant private[variant] (
-    Map()
-  }
+  def validMovesOf2And3(situation: Situation): Map[Pos, List[(String, Move)]] = {
+    def generateMove(orig: Pos, dest: Pos, category: String): Some[(String, Move)] =
```
returning a Some type doesn't match practise within SG/lila. Here you can return just the tuple as its is always a tuple and then decide what you want to do with it where ever it is called (e.g. Some(x), or even x.some)

JamesHeppell commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,17 +102,88 @@ abstract class Variant private[variant] (
+
+    def generateSideMoves(lineOfMarbles: List[Pos], direction: Option[String]): List[(String, Move)] = {
+      // "left" is related to the direction
+      def canLeftSideMove(pos: Pos): Boolean =
```
feels like this can be refactored into one function and then handled lower down, as the code is basically the same, e.g. you could match on the tuple output of sideMovesDirFromDir. this might get refacted anyways while you handle the .get

vincentfrochot committed:
chore: codestyle & reword some tests

JamesHeppell commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,17 +102,88 @@ abstract class Variant private[variant] (
-    Map()
-  }
-    
+    def generateMovesForNeighbours(pos: Pos, neighbour: Pos, direction: Option[String]): List[(String, Move)] = {
```
why pass in direction when it is always the same as pos to neighbour, we could instead calculate it on first line as a val?

vincentfrochot replied:
indeed I'll remove that parameter to just calculate it here

JamesHeppell commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -102,17 +102,88 @@ abstract class Variant private[variant] (
+      val activePlayerPieces = situation.board.piecesOf(situation.player)
+      val opponentPieces = situation.board.piecesOf(!situation.player)
+      val moves = List(
+        neighbour.dir(direction).toList.flatMap {
```
this feels very convoluted, a different approach could be to collect the next 4 squares after the neighbour, which will either be empty or have a piece in them, and then do a match on each of the cases at the same level instead of a nested if/case statement. I think it would read better and be easier to understand? You dont have to change this code if it's functional but its another way to view the problem.

vincentfrochot replied:
oh yes indeed this would be better...

vincentfrochot replied:
I would love to implement it like that if we have enough time : let's see how Abalone progress overall in the next weeks and/or if there is a need to speed up the algos ?

vincentfrochot committed:
chore: codestyle - direction can be computed from 2 Pos

vincentfrochot committed:
chore: refactor side moves

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -46,75 +46,159 @@ abstract class Variant private[variant] (
+      def possibleSideMoves: List[Option[(Pos, Pos)]] = possibleSideMovesDirections.map(
+        dir =>
+          if (lineOfMarbles.flatMap(
+            (pos) => Some(canMoveTowards(pos, dir))
```
because the inside of your `flatMap` can only ever return a `Some` (never `None`), you can remove the addition of `Some` and just use `map`

vincentfrochot replied:
should be ok

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -46,75 +46,159 @@ abstract class Variant private[variant] (
+          if (lineOfMarbles.flatMap(
+            (pos) => Some(canMoveTowards(pos, dir))
+          ).contains(false)) None
+          else List(
```
there's no need to use `List` here, you can just use a tuple and match on the tupe

vincentfrochot replied:
should be ok

vincentfrochot committed:
chore: improve codestyle

- return empty List() instead of None before flattening when we are supposed to
return a List overall
- remove unnecessary use of a List
- use map in case the inner code can only ever return something, e.g. :
  (x) => "a value whatever the value of x"

vincentfrochot committed:
feat: boardAfter

JamesHeppell commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -193,72 +192,86 @@ abstract class Variant private[variant] (
+        1. identify the type of move
+          - line move : no direction to apply (a line move is just 1 marble moving)
+          - push : 1 direction, and dest contains a marble
+          - side move : 2 direction:
```
is the direction not the same for the side move pieces, so why not just pass in one direction? Given you don't want a direction for line move, you can now have Option[String], and then match on this value for your cases, instead of if statements. A direction must match the direction of orig dest if its a push, otherwise its a side move?

vincentfrochot replied:
because I want to be able to find back all the pieces quickly, then apply the direction of the side move to all these

JamesHeppell replied:
ah the directions is storing more info as it also stores if more pieces are moving... great thanks.

JamesHeppell commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -193,72 +192,86 @@ abstract class Variant private[variant] (
+    } else { // side move
+        // from orig, move of the second direction
+        // then move from orig of the first direction and repeat until we have reached dest
+        if ( orig.dir(directions(1)).get.dir(directions(0)).get.index != dest.index ) {
```
I think the issue of using .get here is that each time you are using directions or POS.dir() you are getting back an Option[Pos] and then essentially having to check that its a valid position and/or handle the case for None. Perhaps this can be done at the top of the function or condition and/or using match so that you can use the value more easily?
e.g. 
```
orig.dir(directions(1)) match {
    case Some(p) => <now you can use p as a pos>
    case None => <dont update piece map as this is an error?>
}
```

vincentfrochot committed:
chore: use Directions instead of Option[String]

Also start a refacto of piecesAfterAction - needs more work to be done

vincentfrochot commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -46,75 +46,159 @@ abstract class Variant private[variant] (
-  }
-    
+      List(
+        if (lineOfMarbles.size == 3) generateSideMoves(lineOfMarbles.dropRight(1), direction)
```
I've changed it to : 
```
      List(
        if (lineOfMarbles.size == 3) generateSideMoves(lineOfMarbles.dropRight(1), direction) else List(),
        possibleSideMoves.flatMap {
          case ( (orig, dest, dir) ) => Some(createMove("side", orig, dest, List(dir, direction)))
          case _ => None
        }
      ).flatten
```

vincentfrochot committed:
chore: remove done TODO

vincentfrochot committed:
feat: 3fold, improve codestyle

- trigger draw by 3fold repetition
- delegate to Move.finalizeAfter all the work except the pieceMap update
  which remain in Variant.boardAfter
- improve Situation.piecesAfterAction codestyle
- Replay.replayMove now works
- remove notion of promotion
- add more unit tests

vincentfrochot committed:
chore: codestyle Variant.piecesAfterAction

use match case when working with Option: not using get anymore, when possible

vincentfrochot committed:
fix: Game make use of metrics

vincentfrochot committed:
chore: remove copy pasted false comment

vincentfrochot committed:
fix: get rid of autoEndTurn in Move

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -46,128 +46,233 @@ abstract class Variant private[variant] (
+    val origSideMove: Option[Pos]     = sideDir.flatMap(direction => direction(orig))
+    val destLineMove: Option[Pos]     = lineDir.flatMap(direction => direction(dest))
+    val destLineOneMove: Option[Pos]  = lineDir.flatMap(direction => if (destLineMove.isDefined) direction(destLineMove.get) else None)
+    val destLineTwoMove: Option[Pos]  = lineDir.flatMap(direction => if (destLineOneMove.isDefined) direction(destLineOneMove.get) else None)
```
Use nested flatMaps here

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -46,128 +46,233 @@ abstract class Variant private[variant] (
+          else // ooo
+            pieces + (sideMovePos -> pieces(orig)) - orig +
+              (diagonalPos -> pieces(lineMovePos)) - lineMovePos +
+              (diagonalPos -> pieces(lineMovePos)) - diagonalPos
```
is this correct? the first part of L256 and L257 are doing the same adding of `Pos`->`Piece`?
In general this whole function `piecesAfterAction` is complicated (it has to be) and i havent spent a lot of time to understand each piece of it. If you think it would be beneficial i suggest a call to review it together

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -191,45 +296,36 @@ abstract class Variant private[variant] (
   def valid(@nowarn board: Board, @nowarn strict: Boolean): Boolean = true
 
-  val roles: List[Role] = Role.all
+  def isIrreversible(move: Move): Boolean = move.capture != None
```
are we using `capture` to define when a pieces is pushed off the board and 'scored'? if so this makes sense. Personally I would rewrite as `move.capture.nonEmpty` but it is functional

vincentfrochot replied:
yes we are using capture to define when a pieces is pushed off the board and 'scored'

msomatt commented on src/main/scala/abalone/StartingPosition.scala:
```diff
@@ -20,6 +20,7 @@ case class StartingPosition(
   def initial = fen == format.Forsyth.initial
 }
 
+// @TODO: adapt these
```
That's a long term 'nice to have' TODO. We dont need for initial release

vincentfrochot replied:
agreed

msomatt commented on src/main/scala/abalone/variant/Variant.scala:
```diff
@@ -46,128 +46,233 @@ abstract class Variant private[variant] (
-    Map()
-  }
-    
+  def validMovesOf2And3(situation: Situation): Map[Pos, List[(String, Move)]] = {
```
Is the 'category' part functional? Or was it just for debugging?
As in can we rewrite to get a signature of:
`def validMovesOf2And3(situation: Situation): Map[Pos, List[(Move)]]`?

vincentfrochot committed:
fix: revert autoEndTurn to get it back

- In case we want Abalone to be a multiaction game, it would already be
ready for trying it out
- Also add a few notes for later

vincentfrochot committed:
fix: side moves pieceMap update

- update piecesAfterAction to fix the bug with side moves and remove
  remaining .get()
- add tests specific to side moves
- adapt package name and imports in test files in order to still be able
to use .pp()

vincentfrochot committed:
feat: remove move cat in validMovesOf2And3

was there for debug purpose

vincentfrochot committed:
chore: remove unused dir as String methods

vincentfrochot committed:
feat: update PieceMap based on 2 Pos only

vincentfrochot committed:
feat: replayMove invoke boardAfter

remove Direction from moves creation parameters list in Variant
switch to pair instead of triplet in possibleSideMoves as the direction
is not used anymore later on
